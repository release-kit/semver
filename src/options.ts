import { getInput } from '@actions/core'
import { z } from 'zod'

const CommonSchema = z.object({
  fallback: z.string().min(1).default('v0.1.0'),
  pattern: z
    .string()
    .refine((value) => {
      try {
        void new RegExp(value)
        return true
      } catch {
        return false
      }
    })
    .default('^v?(.*)$'),
})

const TagSourceSchema = CommonSchema.extend({
  source: z
    .union([z.literal('workflow-tag'), z.literal('latest-tag')])
    .default('workflow-tag'),
}).strict()

const StringSourceSchema = CommonSchema.extend({
  source: z.literal('string').default('string'),
  string: z.string().min(1),
}).strict()

const AnySourceSchema = z.union([TagSourceSchema, StringSourceSchema])

interface Inputs {
  source?: string
  string?: string
  fallback?: string
  pattern?: string
}

function getActionInputs(): Inputs {
  return Object.entries({
    source: getInput('source'),
    string: getInput('string'),
    fallback: getInput('fallback'),
    pattern: getInput('pattern'),
  }).reduce<Inputs>((acc: Inputs, [name, value]) => {
    if (!value) return acc
    acc[name as keyof Inputs] = value
    return acc
  }, {})
}

export function getOptions(input: Inputs = getActionInputs()) {
  return AnySourceSchema.parse(input)
}
