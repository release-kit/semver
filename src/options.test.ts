import { getOptions } from './options'

describe('getOptions', () => {
  test('empty inputs', () => {
    expect(getOptions({})).toMatchSnapshot()
  })

  test('unknown source', () => {
    expect(() => getOptions({ source: 'unknown' })).toThrowError()
  })

  test('only string input', () => {
    expect(getOptions({ string: 'v1.0.0' })).toMatchSnapshot()
  })

  test('string input with correct source', () => {
    expect(getOptions({ source: 'string', string: 'v1.0.0' })).toMatchSnapshot()
  })

  test('string input with wrong source', () => {
    expect(() =>
      getOptions({ source: 'workflow-tag', string: 'v1.0.0' })
    ).toThrowError()

    expect(() =>
      getOptions({ source: 'unknown', string: 'v1.0.0' })
    ).toThrowError()
  })

  test('workflow tag source', () => {
    expect(getOptions({ source: 'workflow-tag' })).toMatchSnapshot()
  })

  test('workflow tag source', () => {
    expect(getOptions({ source: 'workflow-tag' })).toMatchSnapshot()
  })

  test('latest tag source', () => {
    expect(getOptions({ source: 'latest-tag' })).toMatchSnapshot()
  })

  test('good pattern', () => {
    expect(getOptions({ pattern: '^.*$' })).toMatchSnapshot()
  })

  test('invalid pattern', () => {
    expect(() => getOptions({ pattern: '(' })).toThrowErrorMatchingSnapshot()
  })

  test('good fallback', () => {
    expect(getOptions({ fallback: 'v1.0.0' })).toMatchSnapshot()
  })

  test('invalid fallback', () => {
    expect(() => getOptions({ fallback: '' })).toThrowErrorMatchingSnapshot()
  })
})
