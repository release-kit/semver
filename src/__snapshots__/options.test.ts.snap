// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getOptions empty inputs 1`] = `
Object {
  "fallback": "v0.1.0",
  "pattern": "^v?(.*)$",
  "source": "workflow-tag",
}
`;

exports[`getOptions good fallback 1`] = `
Object {
  "fallback": "v1.0.0",
  "pattern": "^v?(.*)$",
  "source": "workflow-tag",
}
`;

exports[`getOptions good pattern 1`] = `
Object {
  "fallback": "v0.1.0",
  "pattern": "^.*$",
  "source": "workflow-tag",
}
`;

exports[`getOptions invalid fallback 1`] = `
"[
  {
    \\"code\\": \\"too_small\\",
    \\"minimum\\": 1,
    \\"type\\": \\"string\\",
    \\"inclusive\\": true,
    \\"exact\\": false,
    \\"message\\": \\"String must contain at least 1 character(s)\\",
    \\"path\\": [
      \\"fallback\\"
    ]
  }
]"
`;

exports[`getOptions invalid pattern 1`] = `
"[
  {
    \\"code\\": \\"custom\\",
    \\"message\\": \\"Invalid input\\",
    \\"path\\": [
      \\"pattern\\"
    ]
  }
]"
`;

exports[`getOptions latest tag source 1`] = `
Object {
  "fallback": "v0.1.0",
  "pattern": "^v?(.*)$",
  "source": "latest-tag",
}
`;

exports[`getOptions only string input 1`] = `
Object {
  "fallback": "v0.1.0",
  "pattern": "^v?(.*)$",
  "source": "string",
  "string": "v1.0.0",
}
`;

exports[`getOptions string input with correct source 1`] = `
Object {
  "fallback": "v0.1.0",
  "pattern": "^v?(.*)$",
  "source": "string",
  "string": "v1.0.0",
}
`;

exports[`getOptions workflow tag source 1`] = `
Object {
  "fallback": "v0.1.0",
  "pattern": "^v?(.*)$",
  "source": "workflow-tag",
}
`;

exports[`getOptions workflow tag source 2`] = `
Object {
  "fallback": "v0.1.0",
  "pattern": "^v?(.*)$",
  "source": "workflow-tag",
}
`;
